### ISSUE TITLE
How to extract features vector i.e outputs of the last convolutional layer using the existing weights for the lets x classes from yolov8n.pt pre-train model ?

### ISSUE THREAD

**Issue by TapendraBaduwal** [ultralytics/ultralytics#2398]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

How to extract features vector i.e outputs of the last convolutional layer from yolov8n.pt pre-train model ?

Also i want to a new fully connected layer to the model with the number of nodes equal to the number of new classes we want to add.

Add new class and update nc in .yaml files

And want to Train Model with freezing the weights of pre-trained model.

Help me to guide with code  ?

**Comment by dinhthihuyen** [comment]
@glenn-jocher, I am trying to get image feature vector from YOLOv8 and I do not find file models.py, may be it is in folder of YOLOv5 instead of YOLOv8. I found function test_predict_img() with the following code in file test_python.py in folder /tests:
 # Test tensor inference
    im = cv2.imread(str(SOURCE))  # OpenCV
    t = cv2.resize(im, (32, 32))
    t = torch.from_numpy(t.transpose((2, 0, 1)))
    t = torch.stack([t, t, t, t])
    results = model(t, visualize=True)
Is "results" the image feature vectors, what you mention about?
Thanks so much for your support!

**Comment by glenn-jocher** [comment]
@dinhthihuyen hello! Thank you for reaching out.

Actually, in YOLOv8, you are correct that the `models.py` file does not exist. All model architectures, including YOLOv8, are defined in the `yolo.py` file under the YOLOv8 directory. If you want to modify the model structure, that is the file you should look into.

As for your second question, the `results` object from the `model(t, visualize=True)` call is actually an instance of the `Detect` object, which contains the detection results (bounding boxes, classes, scores) computed by the YOLOv8 model, not the image feature vectors. 

Extracting features directly from an image using YOLOv8 requires accessing the output of the second-to-last layer (before the final prediction layers). This does not happen in the standard forward pass, so it would require modifying the forward method of the model to return these feature vectors.
  
Please note that prior machine learning and deep learning knowledge, particularly familiarity with convolutional neural networks, is essential for making these changes. I hope this clarifies your question, let me know if you need further assistance!

**Comment by dinhthihuyen** [comment]
@glenn-jocher, I trained YOLOv8 detection model for my custom dataset, using pretrained model trained on COCO dataset. Now when I finished training the model, I can detect object in an arbitrary image. I want to get feature map of bounding box detected in image for following problems, such as detailed classification, and I do the following steps: 

 model_detect = YOLO("best.pt")
  SOURCE = "test.jpg"
  im = cv2.imread(str(SOURCE))  
  t = cv2.resize(im, (640, 640))
  t = torch.from_numpy(t.transpose((2, 0, 1)))
  t = torch.stack([t, t, t, t])
  t = t.float()
  results = model_detect.model.forward(t)

I want to ask for your confirmation about my understanding: 
If I want to get feature map of bounding box for another problems, such as detailed classification (as I say before):
+ From "results", is it true if I can get feature map of bounding box with different stride (8, 16, 32) at different stage (4, 6, 9)?
+ As I see, backbone of YOLOv8 object detection and YOLOv8 classification are quite similar, except stage 9 (SPPF). Is it available if I use weight in first 8 stages to match first 8 stages in Classification model (detailed classification problem) and train only head of Classification model?

Thanks so much for your support!!!

**Comment by glenn-jocher** [comment]
Hello @dinhthihuyen!

You're on the right track!

To your first question, indeed, the YOLOv8 object detection model produces feature maps at three different stages with strides of 8, 16, and 32. These stages correspond to the three sequential convolutional layers at the end of the model where detection is performed. So, your understanding is correct.

Regarding your second question, yes, YOLOv8 does share similar architecture for the backbone of both its detection and classification tasks. Correctly understanding that you can utilize this commonality, you can indeed use pretrained weights from the first 8 stages of the YOLOv8 object detection model, and apply them to the corresponding stages of the classification model. Then, you can fine-tune or retrain the "head" of the classification model, and this approach could potentially give you good results due to the transfer learning effect.

Remember though, the effectiveness of adapting pretrained weights from the detection to the classification task, and how much fine-tuning is necessary, would largely depend on how similar your custom dataset is to the original training material (i.e., COCO dataset). 

I hope this answer aids you. Let me know if you have additional inquiries!

**Comment by dinhthihuyen** [comment]
Hello @glenn-jocher,
I got feature map of image and coordinate of bounding box extracted from image. Now I want to get feature map of bounding box extracted using its coordinates and feature map of the whole image. I read and find out Roi Align can help me do it, but it is quite difficult to implement. I want to ask you if YOLOv8 has some functions to do it? Or could you recommend other methods I could use to get it completed?
Thanks so much for your support!

**Comment by glenn-jocher** [comment]
@dinhthihuyen hello! In YOLOv8, there isn't a built-in utility or function that directly performs ROI Align on the feature maps to extract the features corresponding to the bounding box regions. The YOLO series primarily focuses on detection task, so the functionality provided pertains to that. 

However, you're correct in identifying ROI Align as an appropriate method to get the features of a bounding box from the feature map of the whole image. The ROI Align technique, which comes from Mask-RCNN, helps to convert the features within an arbitrary bounding box into a fixed-size feature map.

Implementation wise, if you're using PyTorch, a popular deep learning library, torchvision.ops contains a RoIAlign layer which is quite straightforward to use and could serve your purpose. Another option can be to use a library like detectron2 from Facebook AI, which has comprehensive mask-rcnn implementations including RoI layers.

Just remember that the end results of using ROI Align could depend upon the kind of problem you are dealing with. Always ensure that the subsequent layers of your model are capable enough of extracting useful information out of the RoI aligned features.

I hope this information assists you in accomplishing your goal, and don't hesitate to ask if you have any further queries!

**Comment by Kai-054** [comment]
@dinhthihuyen Looks like your problem has been solved! 
I'm having the same problem as yours. 
I want to find the output vector after the data is forwarded over the network,
but I still haven't found "that vector output for them to take into account the loss"
can you help me?
Thank you very much

**Comment by dinhthihuyen** [comment]
As I understand, if you use model yolov8.pt to detect object, when you call  `model(t, visualize=True)`  to detect result, it will go through function `predict` (nn/task.py), and **visualize** equals **True** enable you to save feature map at every layers of model.
If you want to get output vector after data is forwarded over the network, and you use model yolov8 (as config in cfg/models/v8/yolov8.yaml), features of data would be extracted at 3 layers 15, 18 and 21  - feature maps at three different stages with strides of 8, 16, and 32 that model YOLOv8 produced. So, instead of saving feature maps at every layers of model, you can concentrate on exactly feature map at layers 15, 18, 21, which I think is what you need.
